jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+            baseResponse.getHttpFields().add(ConnectionUpgrade);
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_ACCEPT, AcceptHash.hashKey(key));
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_EXTENSIONS,
+                                             ExtensionConfig.toHeaderValue(connection.getExtensionStack().getNegotiatedExtensions()));
+            if (subprotocol!=null)
+                baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL,subprotocol);
+            baseResponse.flushBuffer();
+            baseRequest.setHandled(true);
+
+            // upgrade
+            if (LOG.isDebugEnabled())
+                LOG.debug("upgrade connection={} session={}",connection,session);
+
+            baseRequest.setAttribute(HttpConnection.UPGRADE_CONNECTION_ATTRIBUTE, connection);
+            return true;
+        }
+    };
+
+
+
+    static class TestWebSocketConnectionFactory extends ContainerLifeCycle implements WebSocketConnectionFactory
+    {
+        WebSocketPolicy policy = new WebSocketPolicy(WebSocketBehavior.SERVER);
+        ByteBufferPool bufferPool = new MappedByteBufferPool();
+        WebSocketExtensionRegistry extensionRegistry = new WebSocketExtensionRegistry();
+        DecoratedObjectFactory objectFactory = new DecoratedObjectFactory();
+
+        @Override
+        public WebSocketCoreConnection newConnection(Connector connector, EndPoint endPoint, List<String> extensions)
+        {
+            ExtensionStack extensionStack = new ExtensionStack(extensionRegistry);
+            extensionStack.negotiate(objectFactory, policy, bufferPool,
+                                     extensions.stream().map(ExtensionConfig::parse).collect(Collectors.toList()));
+
+            UpgradeRequest upgradeRequest = null; // TODO
+            UpgradeResponse upgradeResponse = null; // TODO
+
+            return new WebSocketCoreConnection(
+                    endPoint,
+                    connector.getExecutor(),
+                    bufferPool,
+                    objectFactory,
+                    policy,
+                    extensionStack,
+                    upgradeRequest,
+                    upgradeResponse);
+        }
+    }
+
+
+    static class TestWebSocketSessionFactory extends ContainerLifeCycle implements WebSocketSessionFactory
+    {
+        @Override
+        public WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint>
+        newSession(WebSocketCoreConnection connection, String subprotocol)
+        {
+
+            // TODO why is remoteEndpoint an interface rather than just an impl?
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
+            remoteEndpoint.open(); // TODO why is this needed?
+
+            // TODO abstract the creation of a local Endpoint
+            WebSocketLocalEndpoint localEndpoint = new WebSocketLocalEndpoint()
```

Comment:
You don't create a WebSocketLocalEndpoint, that's not a User API.

In a server usage, use the WebSocketCreator to obtain an API specific endpoint instance, then a factory to create/manage the metadata for that instance, and the factory is then capable of providing WebSocketLocalEndpoint instances from the metadata that is bound to that specific instance.